### ISSUE TITLE
unsupported-binary-operation is reported for type annotations

### ISSUE THREAD

**Issue by sam-s** [PyCQA/pylint#8297]
### Bug description

The following code works fine and conforms to [examples](https://docs.python.org/3/library/typing.html):
```python
"https://github.com/PyCQA/pylint/issues/8297"
from dataclasses import dataclass

@dataclass
class C:
    "class doc"
    a: str
    b: int = 10
    c: float | int | None = None
```

### Command used

```shell
pylint a.py
```

### Pylint output

```shell
************* Module a
a.py:9: [E1131(unsupported-binary-operation), C] unsupported operand type(s) for |
a.py:9: [E1131(unsupported-binary-operation), C] unsupported operand type(s) for |
```

### Expected behavior

no diagnostics

### Pylint version

```shell
pylint 2.15.10
astroid 2.13.5
Python 3.10.6 (main, Nov 14 2022, 16:10:14) [GCC 11.3.0]
```

### OS / Environment

ubuntu

**Comment by Pierre-Sassoulas** [comment]
I can reproduce on 2.17.0dev0, (only one message instead of two).

**Comment by mbyrnepr2** [comment]
I think its a duplicate of https://github.com/PyCQA/pylint/issues/7381. There are multiple examples of this happening over there and I suspect it has a similar root-cause.
